### ISSUE TITLE
Incorrect generic type inference for wildcards

### ISSUE THREAD

**Issue by atomicknight** [typetools/checker-framework#511]
Environment: Checker Framework 1.9.7, Oracle JDK 1.8.0_65, Windows 7

A third-party library (ANTLR) defines an API that uses wildcards in the method signature; users of the library override the method. The standard Oracle javac accepts this without issue. However, the Checker Framework raises an error, seemingly due to incorrect type inference for the wildcards.

The method signature looks like this:
```java
void syntaxError(Recognizer<?, ?> recognizer, Object offendingSymbol, int line, int charPositionInLine, String msg, RecognitionException e)
```

The error that is raised looks like this:
```
[override.param.invalid] void syntaxError(Test this, Recognizer<? extends Object, ? extends ATNSimulator> p0, Object p1, int p2, int p3, String p4, RecognitionException p5) in org.example.Test cannot override void syntaxError(ANTLRErrorListener this, Recognizer<? extends Object, ? extends Object> p0, Object p1, int p2, int p3, String p4, RecognitionException p5)
```
Note the incorrect bound for the second type parameter of Recognizer.

The full test case (Maven build - just execute `mvn`): https://github.com/atomicknight/cf-testcases/tree/master/genericwildcardinheritance1

**Comment by mernst** [comment]
Thanks for the bug report.  I have committed a test case in directory checker-framework/checker/tests/nullness-genericwildcard/.

**Comment by smillst** [comment]
The issue here is that the Checker Framework does not support wildcards that are outside of the bounds of the type variable.  For example:
```
class MyGeneric<T extends Object>{}
class Use{
   MyGeneric<? extends  Object> wildCardExtendsObject;
   MyGeneric<? extends Number> wildCardExtendsNumber = wildCardExtendsObject;
}
```
Results in this error:
```
:4: error: [assignment.type.incompatible] incompatible types in assignment.
    MyGeneric<? extends Number> wildCardExtendsNumber = wildCardExtendsObject;
                                                        ^
  found   : MyGeneric<? extends Object>
  required: MyGeneric<? extends Number>
```
However, this assignment is legal because the type of wildCardExtendsObject in javac is:
    CAP#1 extends Number from capture of ? extends Object
So this assignment is legal.
